fling/component/group/fedamp_group.py

```diff
@@ -0,0 +1,105 @@
+        self.sigma = args.learn.sigma
+        self.lamda = args.learn.lamda
+        self.client_ws = [None for i in range(self.args.client.client_num)]    # maintain all clients' personalized models
+        self.client_us = [None for i in range(self.args.client.client_num)]    # aggregated model for each client
+
+    def sync(self) -> None:
+        r"""
+        Overview:
+            Send customized global models to each client
+        """
+        if self.client_us[0] is None:
+            super().sync()  # Called during system initialization
+
+        else:
+            for idx, client in enumerate(self.clients):
+                client.client_u = copy.deepcopy(self.client_us[idx])
+
+    def initialize(self) -> None:
+        super().initialize()
+        self.client_ws = [copy.deepcopy(self.clients[i].model) for i in range(self.args.client.client_num)]
+        self.client_us = [copy.deepcopy(self.clients[i].model) for i in range(self.args.client.client_num)]
+
+
+    def receive_models(self):
+        r"""
+        Overview:
+            Receive personalized models from each client
+        """
+        for idx, client in enumerate(self.clients):
+            self.client_ws[idx] = copy.deepcopy(client.model)
+
+    def aggregate(self, train_round: int) -> int:
+        r"""
+        Overview:
+            Aggregate customized global models and send to each client
+        """
+        # recieve models from clients
+        self.receive_models()
+        for i in range(self.args.client.client_num):
+            self.client_ws[i].to(self.args.learn.device)
+            self.client_us[i].to(self.args.learn.device)
+
+        # aggregate models
+        weights = [weight_flatten(mw) for mw in self.client_ws]
+        for i, mu in enumerate(self.client_us):  # calculate u for each client
+            for param in mu.parameters():  # set zero for each parameter
+                param.data = torch.zeros_like(param.data)
+
+            coef = torch.zeros(self.args.client.client_num)
+            for j, mw in enumerate(self.client_ws):
+                if i == j: continue
+                sub = weights[i] - weights[j]
+                sub = torch.dot(sub, sub)
+                coef[j] = self.args.learn.alphaK * self.e(sub)
+            coef[i] = 1 - torch.sum(coef)
+
+            for j, mw in enumerate(self.client_ws):
+                for param, param_j in zip(mu.parameters(), mw.parameters()):
+                    param.data += coef[j] * param_j
+
+        for i in range(self.args.client.client_num):
+            self.client_ws[i].to('cpu')
+            self.client_us[i].to('cpu')
+
+        # send to all clients
+        self.sync()
+
+        # perform alphaK decay
+        if train_round % self.args.learn.decay_frequency == 0 and train_round != 0:
+            self.args.learn.alphaK *= self.args.learn.decay_rate
+
+        # calculate communication cost
+        trans_cost = 0
+        state_dict = self.clients[0].model.state_dict()
```

Comment:
The transition cost is not correctly calculated. Since you use copy.deepcopy, you exactly transfer the total model in each training round. So, I think you should only perform fedamp aggregation on ``fed_keys``.